documentation/http_api_index.markdown

```diff
@@ -0,0 +1,141 @@
+
+Starting with Puppet Server 2.1, Server provides both the current and previous API endpoints, and can serve nodes running Puppet agent 3.x and 4.x. However, Rack masters, WEBrick masters, and Puppet Server 2.0 cannot serve nodes running Puppet 3.x.
+
+## Puppet and Puppet CA APIs
+
+Beginning with Puppet 4, Puppet's HTTP API was split into two separately versioned APIs:
+
+-   An API for configuration-related services
+-   An API for the certificate authority (CA).
+
+All configuration endpoints are prefixed with `/puppet`, while all CA endpoints are prefixed with `/puppet-ca`. All endpoints are explicitly versioned: the prefix is always immediately followed by a string like `/v3` (a directory separator, the letter `v`, and the version number of the API).
+
+### Authorization
+
+Authorization for `/puppet` endpoints is still controlled with Puppet's `auth.conf` authorization system.
+
+Puppet Server ignores `auth.conf` for `/puppet-ca` endpoints. Access to the `certificate_status` endpoint is configured in Puppet Server's `ca.conf` file, and the remaining CA endpoints are always accessible. Rack Puppet master servers still use `auth.conf` for `/puppet-ca`.
+
+When specifying authorization in `auth.conf`, the prefix, and the version number (such as `/puppet/v3`) on the paths must be retained, since Puppet matches authorization rules against the full request path.
+
+## Puppet V3 HTTP API
+
+The Puppet agent application uses several network services to manage systems. These services are all grouped under the `/puppet` API. Other tools can access these services and use the Puppet master's data for other purposes.
+
+The V3 API contains endpoints of two types: those that are based on dispatching to Puppet's internal "indirector" framework, and those that are not (namely the [environment endpoints](#environment-endpoints)).
+
+Every HTTP endpoint that dispatches to the indirector follows the form `/puppet/v3/:indirection/:key?environment=:environment`, where:
+
+-   `:environment` is the name of the environment that should be in effect for the request. Not all endpoints need an environment, but the query parameter must always be specified.
+-   `:indirection` is the indirection to which the request is dispatched.
+-   `:key` is the "key" portion of the indirection call.
+
+Using this API requires significant understanding of how Puppet's internal services are structured, but the following documents specify what is available and how to interact with it.
+
+### Configuration management services
+
+The Puppet agent application directly uses these servcies to manage the configuration of a node.
+
+These endpoints accept payload formats formatted as JSON or PSON (MIME types of `application/json` and `text/pson`, respectively) except for `File Content` and `File Bucket File`, which always use `application/octet-stream`.
```

Comment:
JSON is the default, and while I don't think we've formally deprecated it we're looking to replace the remaining usages of PSON in Puppet 6 (unless users have binary content they should not be using PSON, and generally they should only be receiving binary content via the `File Content` and `File Bucket File` endpoints. I'd appreciate it if we didn't put JSON and PSON on the same footing going forward and caveated it's usage as fallback for binary content that we hope to provide a better solution for soon (this is the Rich Data Types feature).